packages/jsii/lib/assembler.ts

```diff
@@ -0,0 +1,879 @@
+import colors = require('colors/safe');
+import crypto = require('crypto');
+import deepEqual = require('deep-equal');
+import fs = require('fs-extra');
+import spec = require('jsii-spec');
+import log4js = require('log4js');
+import path = require('path');
+import ts = require('typescript');
+import { Diagnostic, Emitter } from './emitter';
+import literate = require('./literate');
+import { ProjectInfo } from './project-info';
+import utils = require('./utils');
+import { Validator } from './validator';
+
+// tslint:disable:no-var-requires Modules without TypeScript definitions
+const sortJson = require('sort-json');
+// tslint:enable:no-var-requires
+
+const LOG = log4js.getLogger('jsii/assembler');
+
+/**
+ * The JSII Assembler consumes a ``ts.Program`` instance and emits a JSII assembly.
+ */
+export class Assembler implements Emitter {
+    private _diagnostics = new Array<Diagnostic>();
+    private _deferred = new Array<() => void>();
+    private _types: { [fqn: string]: spec.Type };
+
+    /**
+     * @param projectInfo information about the package being assembled
+     * @param program     the TypeScript program to be assembled from
+     */
+    public constructor(public readonly projectInfo: ProjectInfo,
+                       public readonly program: ts.Program) {}
+
+    private get _typeChecker(): ts.TypeChecker {
+        return this.program.getTypeChecker();
+    }
+
+    /**
+     * Attempt emitting the JSII assembly for the program.
+     *
+     * @return the result of the assembly emission.
+     */
+    public async emit(): Promise<EmitResult> {
+        this._diagnostics = [];
+        if (!this.projectInfo.description) {
```

Comment:
Wouldn't it make more sense to perform these validations in `ProjectInfo`?

Reply:
I decided to have `ProjectInfo` remain rather un-opinionated, especially w/r/t optionals. Could make it also emit `diagnostics`, I suppose. I'm going to hold off on this one for now though.